directives/animate/animate.js

```diff
@@ -0,0 +1,273 @@
+				)
+			);
+		}, timeout);
+	}
+
+	async getElemInfo() {
+
+		await new Promise((r) => requestAnimationFrame(r));
+		const style = window.getComputedStyle(this.elem);
+
+		const styleAttr = this.elem.getAttribute('style');
+		if (style.display !== 'flex') {
+			this.elem.style.display = 'grid'; // forces margins to collapse during size calculation
+		}
+
+		const rect = this.elem.getBoundingClientRect();
+
+		let top = 0;
+		let left = 0;
+		const offsetParent = this.elem.offsetParent;
+		let e = this.elem;
+		while (e && e !== document && !(e instanceof DocumentFragment)) {
+			if (e === offsetParent) {
+				break;
+			}
+			top += e.offsetTop - (e.scrollTop || 0);
+			left += e.offsetLeft - (e.scrollLeft || 0);
+			e = getComposedParent(e);
+		}
+
+		this.elem.style.removeProperty('display');
+
+		const marginsH = (parseInt(style.marginLeft) || 0) + (parseInt(style.marginRight) || 0);
+		const marginsV = (parseInt(style.marginTop) || 0) + (parseInt(style.marginBottom) || 0);
+
+		let cloneHeight = 0;
+		if (this.clone !== null) {
+			const cloneRect = this.clone.getBoundingClientRect();
+			cloneHeight = cloneRect.height;
+		}
+
+		return {
+			clone: {
+				height: cloneHeight,
+				fullHeight: rect.height + marginsV,
+				width: rect.width + marginsH
+			},
+			elem: {
+				height: rect.height,
+				left,
+				opacity: this.state === 'showing' && this.clone === null ? '0' : style.opacity,
+				styleAttr,
+				top,
+				width: rect.width
+			}
+		};
+
+	}
+
+	async hide(opts) {
+
+		if (this.state === 'hiding' || this.state === 'hidden') {
+			return;
+		}
+		this.state = 'hiding';
+
+		// if focus is inside and was placed by keyboard, move it to next focusable
+		const activeElem = getComposedActiveElement();
+		if (activeElem && activeElem.classList.contains('focus-visible')) {
+			const focusIsInside = isComposedAncestor(this.elem, activeElem);
+			if (focusIsInside) {
+				const nextFocusable = getNextFocusable(activeElem);
+				if (nextFocusable) nextFocusable.focus();
+			}
+		}
+
+		if (reduceMotion || opts.skip === true) {
+			this.state = 'hidden';
+			this.elem.setAttribute('hidden', '');
+			this.dispatchEvent(100);
+			return;
+		}
+
+		const info = await this.getElemInfo();
+
+		return this.animate({
+			clone: {
+				start: {
+					height: `${info.clone.height || info.clone.fullHeight}px`,
+					transition: `height ${hideTransitionDuration}ms ease-out`,
+					width: `${info.clone.width}px`
+				},
+				end: {
+					height: '0'
+				}
+			},
+			elem: {
+				start: {
+					height: `${info.elem.height}px`,
+					left: `${info.elem.left}px`,
+					opacity: info.elem.opacity,
+					position: 'absolute',
+					top: `${info.elem.top}px`,
+					transitionProperty: 'height, opacity, transform',
+					transitionDuration: `${hideTransitionDuration}ms`,
+					transitionTimingFunction: 'ease-out',
+					width: `${info.elem.width}px`
+				},
+				end: {
+					opacity: '0',
+					transform: `translateY(-${moveYValue}px)`
```

Comment:
Would it be as simple as `translate${direction}(...` to make this work horizontally as well? I'm thinking about carousels, if that's the right term, like this one in Wave, but probably plenty of use cases.

![animate-x](https://user-images.githubusercontent.com/1289042/110949191-4f6a1600-8310-11eb-85f8-a44fd083fdf4.gif)

Reply:
I suppose that would require the ability to specify the value as negative or positive explicitly as well.

Reply:
... and possibly events if you wanted to start the fade-in of the next element after the fade-out of the previous element was done.

Reply:
This directive is only complex due to the craziness we have to do to "slide" the thing in or out while growing/shrinking the clone to fill its space while that's happening.

If all we needed was a simple translation in one direction, I'm not sure you'd need a directive? I'd implement that carousel just using a container that doesn't allow overflow and then when the user arrows translate everything inside it left/right, which could be animated.

Reply:
Oooh, that makes sense. Ok, so the carousel is a bad example, but I wonder if there would be a need to do it horizontally. I suppose we can cross that bridge when we get to it.

Reply:
Yeah totally. We could easily extend this in the future by either adding more options to these existing directives (e.g. `show({direction: "right"})`) or yeah by introducing new ones like `slide-in` or whatever.

I think for most animations it should be easy enough for teams to just do them using regular CSS transitions in their components. But for cases where there's a complex series of transitions like this where we want to centralize that logic, creating new animation directives could make sense.